src/HealthChecks.AzureStorage/AzureFileShareHealthCheck.cs

```diff
@@ -6,26 +6,39 @@ namespace HealthChecks.AzureStorage
 {
     public class AzureFileShareHealthCheck : IHealthCheck
     {
-        private readonly string _connectionString;
-        private readonly string? _shareName;
+        private readonly Func<ShareClient> _clientFactory;
 
         private static readonly ConcurrentDictionary<string, ShareClient> _shareClientsHolder = new();
 
         public AzureFileShareHealthCheck(string connectionString, string? shareName = default)
         {
-            _connectionString = connectionString ?? throw new ArgumentNullException(nameof(connectionString));
-            _shareName = shareName;
+            if (connectionString == null)
+            {
+                throw new ArgumentNullException(nameof(connectionString));
+            }
+
+            _clientFactory = () => _shareClientsHolder.GetOrAdd($"{connectionString}{shareName}", _ => new ShareClient(connectionString, shareName));
+        }
+
+        public AzureFileShareHealthCheck(ShareClient shareClient)
```

Comment:
I suggest to add one more `AzureFileShareHealthCheck` ctor with `Func<ShareClient>` argument and overloads to  `AzureStorageHealthCheckBuilderExtensions` with `ShareClient` and `IServiceProvider`.

Reply:
I've changed this such that each health check simply has a member representing the service client. How do you feel about this instead of the factories? Was there a reason I missed for resolving the clients within the method as opposed to the ctor? To accommodate the change, I did have to update the unit tests with real connection strings that could be parsed.

If that's ok, do we need the static client caching?

Reply:
> How do you feel about this instead of the factories? 

Fine.

> Was there a reason I missed for resolving the clients within the method as opposed to the ctor?

I think no. In fact, this is even better, because it allows you to identify configuration errors before actual HC execution.

Reply:
> If that's ok, do we need the static client caching?

How will it work without caching?

Reply:
@wsugarman I'm ready to merge this PR but we have the last unresolved conversation here.

Reply:
Ah, sorry! I think alternatively we could just create a client in the registration factory and pass it to the ctor. However, the Azure Storage docs do seem to indicate that singleton clients should be used, so caching the client aligns with that guidance. Furthermore, it is already the behavior today, and I think removing it would be too much of a behavioral change.

So we should probably keep the caching haha

Reply:
@sungam3r I think that's everything